### PR TITLE
fix: branch stmts

### PR DESCRIPTION
**Pull request by petar-dambovaliev**
Adds validation for `break`, `continue` and `fallthrough` to disallow invalid code.
Moves existing validation from the runtime to the preprocessor.

Closes https://github.com/gnolang/gno/issues/2973

### CODE REVIEW

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -2125,8 +2130,19 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
 
 			// TRANS_LEAVE -----------------------
 			case *BranchStmt:
+
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2133]: delete this line

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -2139,6 +2155,8 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
 						}
 					}
 				case CONTINUE:
+					notAllowedFunc("continue")
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2158]: should this logic be merged into `findContinuableNode` and `findBranchLabel`, seems what we need to do is to make these two func correct.

File: gnovm/pkg/gnolang/preprocess.go [hunk 3]
```diff
@@ -2154,17 +2172,36 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
...
-							}
+					swchC, ok := last.(*SwitchClauseStmt)
+					if !ok {
+						// fallthrough is only allowed in a switch statement
+						panic("fallthrough statement out of place")
+					}
+
+					if n.GetAttribute(ATTR_LAST_BLOCK_STMT) != true {
+						// no more clause after the one executed, this is not allowed
+						panic("fallthrough statement out of place")
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2183]: any test targeting this?

File: gnovm/pkg/gnolang/nodes.go [hunk 4]
```diff
@@ -149,16 +149,17 @@ func (loc Location) IsZero() bool {
...
+	ATTR_PREDEFINED      GnoAttribute = "ATTR_PREDEFINED"
+	ATTR_TYPE_VALUE      GnoAttribute = "ATTR_TYPE_VALUE"
+	ATTR_TYPEOF_VALUE    GnoAttribute = "ATTR_TYPEOF_VALUE"
+	ATTR_IOTA            GnoAttribute = "ATTR_IOTA"
+	ATTR_LOCATIONED      GnoAttribute = "ATTR_LOCATIONE"     // XXX DELETE
+	ATTR_GOTOLOOP_STMT   GnoAttribute = "ATTR_GOTOLOOP_STMT" // XXX delete?
+	ATTR_LOOP_DEFINES    GnoAttribute = "ATTR_LOOP_DEFINES"  // []Name defined within loops.
+	ATTR_LOOP_USES       GnoAttribute = "ATTR_LOOP_USES"     // []Name loop defines actually used.
+	ATTR_SHIFT_RHS       GnoAttribute = "ATTR_SHIFT_RHS"
+	ATTR_LAST_BLOCK_STMT GnoAttribute = "ATTR_LAST_BLOCK_STMT"
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/nodes.go:162]: is there a better name?

File: gnovm/pkg/gnolang/preprocess.go [hunk 5]
```diff
@@ -2125,8 +2130,19 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
...
+					_, isFunc := last.(*FuncLitExpr)
+
+					if isFunc {
+						panic(fmt.Sprintf("%s statement out of place", s))
+					}
+				}
+
 				switch n.Op {
 				case BREAK:
+					notAllowedFunc("break")
```
Comment (mvertes) [inline gnovm/pkg/gnolang/preprocess.go:2144]: Your fix seems incomplete. Consider the following case where the bug is not catched:

```go
package main

func main() {
    for i := 0; i < 10; i++ {
        if i == 1 {
            _ = func() int {
                if true {
                    break
                }
                return 11
            }()
        }
        println(i)
    }
    println("wat???")
}

// Error:
// main/files/for24.gno:8:17: break statement out of place
```

File: gnovm/pkg/gnolang/nodes.go (hunk 4 above)
Comment (petar-dambovaliev) [reply gnovm/pkg/gnolang/nodes.go:162]: I am open to suggestions

File: gnovm/pkg/gnolang/preprocess.go (hunk 2 above)
Comment (petar-dambovaliev) [reply gnovm/pkg/gnolang/preprocess.go:2176]: You mean `findBreakableNode`, not `findBranchLabel`.

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (petar-dambovaliev) [reply gnovm/pkg/gnolang/preprocess.go:2175]: There are several tests. grep for `fallthrough statement out of place`

**Review by mvertes — APPROVED** [review]
looks good to me

**Pull request merged**